feat(config)!: experimental flags (#28187)
renovatebot/renovate: docs/usage/self-hosted-experimental-flags.md

Example usage:

```js
experimentalFlags: ['noMavenPomCheck'];
```

## `nugetDownloadNupkgs`

If added to the `experimentalFlags` list, Renovate will download `nupkg` files for determining package metadata.

Example usage:

```js
experimentalFlags: ['nugetDownloadNupkgs'];
```

## `paginateAll`

If added to the `experimentalFlags` list, Renovate will always paginate requests to GitHub fully, instead of stopping after 10 pages.

Example usage:

```js
experimentalFlags: ['paginateAll'];
```

## `rebasePaginationLinks`

If added to the `experimentalFlags` list, Renovate will rewrite GitHub Enterprise Server's pagination responses to use the `endpoint` URL from the Renovate config.

Example usage:

```js
experimentalFlags: ['rebasePaginationLinks'];
```

<!-- prettier-ignore -->
!!! note
    For the GitHub Enterprise Server platform only.

## `repoCacheForceLocal`

If added to the `experimentalFlags` list, Renovate will persist repository cache locally after uploading to S3.

Example usage:

```js
experimentalFlags: ['repoCacheForceLocal'];
```

## `sqlitePackageCache`

If added to the `experimentalFlags` list, Renovate will use SQLite as the backend for the package cache.

> Don't combine with `redisUrl`, Redis would be preferred over SQlite.

Reviewer: ```suggestion
Do not combine with `redisUrl`, as Renovate would then prefer Redis over SQlite.
```

Please check if this is correct. We should expand contractions like `don't` into `do not`. Also `Redis would be preferred over SQlite.` is vague about who prefers this.

I think we're trying to say something like:

> If you set `redisUrl` then Renovate ignores the `sqlitePackageCache` option, because it will use Redis instead of SQLite.

> # Self-hosted experimental flags

Reviewer 2: @rarkins should we add it to the TOC?

Author: What is the TOC? :)

The following flags are "experimental" because they:

- are not commonly needed
- are typically an effort to work around some other service's or platform's problem
- can be removed at any time
- are variables for Renovate's internal use to validate they work as intended

Experimental variables which are commonly used and for which there is no external solution in sight can be converted to an official configuration option by the Renovate bot developers.

Use these experimental flags at your own risk.
We do not follow Semantic Versioning for any experimental variables.
These flags may be removed or have their behavior changed in **any** version.
We will try to keep breakage to a minimum, but make no guarantees that an experimental flag will keep working.

## `dockerHubTags`

> If added to the experimentalFlags list, Renovate will use the Docker Hub API (`https://hub.docker.com`) to fetch tags instead of the normal Docker API for images pulled from `https://index.docker.io`.

Reviewer: ```suggestion
If added to the `experimentalFlags` list, Renovate will use the Docker Hub API (`https://hub.docker.com`) to fetch tags instead of the normal Docker API for images pulled from `https://index.docker.io`.
```

```js
experimentalFlags: ['dockerHubTags'];
```

## `dockerMaxPages`

> If set to an integer, Renovate will use this as max page number for docker tags lookup on docker registries, instead of the default 20 pages.

Reviewer: ```suggestion
The default maximum page number for Docker tags lookup on Docker registries is `20`.
You can override this default by setting a new integer number in the `dockerMaxPages` field.
```

> This is useful for registries which ignores the `n` parameter in the query string and only return 50 tags per page.

Reviewer: ```suggestion
This is useful for registries which ignore the `n` parameter in the query string and only return `50` tags per page.
```

# Self-hosted experimental flags

The following flags are "experimental" because they:

- are not commonly needed
- are typically an effort to work around some other service's or platform's problem
- can be removed at any time
- are variables for Renovate's internal use to validate they work as intended

Experimental variables which are commonly used and for which there is no external solution in sight can be converted to an official configuration option by the Renovate bot developers.

Use these experimental flags at your own risk.
We do not follow Semantic Versioning for any experimental variables.
These flags may be removed or have their behavior changed in **any** version.
We will try to keep breakage to a minimum, but make no guarantees that an experimental flag will keep working.

## `dockerHubTags`

If added to the `experimentalFlags` list, Renovate will use the Docker Hub API (`https://hub.docker.com`) to fetch tags instead of the normal Docker API for images pulled from `https://index.docker.io`.

```js

> experimentalFlags: ['dockerHubTags'];

Reviewer 2: ```suggestion
experimentalFlags: ['dockerHubTags=true'];
```

Shouldn't we explicit use bool values? because we will invert the flag in one of the next majors?

Author: Okay, will update.

> If multiple `autodiscoverTopics` are used resulting order will be per topic not global.

Allowed values:

- `asc`
- `desc`

Default value: `asc`.

Example usage:

```js
experimentalFlags: ['autoDiscoverRepoOrder=desc'];
```

## `dockerHubTags`

If added to the `experimentalFlags` list, Renovate will use the Docker Hub API (`https://hub.docker.com`) to fetch tags instead of the normal Docker API for images pulled from `https://index.docker.io`.

Example usage:

```js
experimentalFlags: ['dockerHubTags'];
```

## `dockerMaxPages`

The default maximum page number for Docker tags lookup on Docker registries is `20`.
You can override this default by setting a new integer number in the `dockerMaxPages` field.
This is useful for registries which ignore the `n` parameter in the query string and only return `50` tags per page.

Example usage:

```js
experimentalFlags: ['dockerMaxPages=10'];
```

## `mergeConfidenceSupportedDatasources`

If set, Renovate will query the merge-confidence JSON API only for datasources that are part of this list.
The expected value for this environment variable is a list of strings separated by a comma `,`.

Example usage:

```js

> experimentalFlags: ['mergeConfidenceSupportedDatasources=[docker,deno]'];

Author: ```suggestion
experimentalFlags: ['mergeConfidenceSupportedDatasources=["docker","deno"]'];
```

# Self-hosted experimental flags

The following flags are "experimental" because they:

- are not commonly needed
- are typically an effort to work around some other service's or platform's problem
- can be removed at any time
- are flags for Renovate's internal use to validate they work as intended

Experimental flags which are commonly used and for which there is no external solution in sight can be converted to an official configuration option by the Renovate bot developers.

Use these experimental flags at your own risk.
These flags may be removed or have their behavior changed in **any** version.
We will try to keep breakage to a minimum, but make no guarantees that an experimental flag will keep working.

> ## `disableDockerHubTags`

Reviewer 2: Add a docs test to validate sort order

Allowed values:

- `alpha`
- `created`
- `updated`
- `size`
- `id`

Default value: `alpha`.

Example usage:

```js
experimentalFlags: ['autoDiscoverRepoSort=alpha'];
```

## `dockerHubTags`

If added to the `experimentalFlags` list, Renovate will use the Docker Hub API (`https://hub.docker.com`) to fetch tags instead of the normal Docker API for images pulled from `https://index.docker.io`.

Example usage:

```js
experimentalFlags: ['dockerHubTags'];
```

## `dockerMaxPages`

The default maximum page number for Docker tags lookup on Docker registries is `20`.
You can override this default by setting a new integer number in the `dockerMaxPages` field.
This is useful for registries which ignore the `n` parameter in the query string and only return `50` tags per page.

Example usage:

```js
experimentalFlags: ['dockerMaxPages=10'];
```

## `mergeConfidenceSupportedDatasources`

If set, Renovate will query the merge-confidence JSON API only for datasources that are part of this list.
The expected value for this environment variable is a list of strings separated by a comma `,`.

Example usage:

```js

> experimentalFlags: ['mergeConfidenceSupportedDatasources=["docker","deno"]'];

Reviewer 3: ```suggestion
experimentalFlags: ['mergeConfidenceSupportedDatasources=docker,deno'];
```

Is there anything stopping this?

Author: There are  places where we have error handling for the json string arrays and I wanted to reuse them instead of modifying. Nothing other than that.

Reviewer 3: I'd prefer to go with my suggestion then as it's simpler
